### PR TITLE
Add semantic class resolution for multi-dataset training

### PR DESCRIPTION
**Pull request by MohanadMahran**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds semantic class resolution for multi-dataset training, automatically aligning class IDs by class names while preserving original annotations and caches.

### 📊 Key Changes
- Introduces `ClassRegistry` for deterministic global class discovery, user-defined ordering, validation, and local-to-global label remapping.
- Extends dataset YAML parsing and dataloader construction to support multiple datasets, including dictionary-based entries with optional metadata.
- Applies runtime remapping across detection and instance segmentation datasets without modifying label files.
- Adds comprehensive documentation, navigation updates, unit tests, and end-to-end training coverage.

### 🎯 Purpose & Impact
- Enables reliable joint training when datasets use different class orders or partially overlapping class sets.
- Preserves backward compatibility for existing single-dataset workflows and reusable dataset caches.
- Improves usability by removing the need for manual annotation rewriting or class-ID mapping tables.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @MohanadMahran, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message helps with the initial review; an Ultralytics engineer will assist with your contribution. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/data/build.py [hunk 1]
```diff
@@ -244,6 +245,70 @@ def build_yolo_dataset(
...
+                    pad=pad,
+                    prefix=colorstr(f"{mode}: (ds{i}) "),
+                    task=cfg.task,
+                    classes=cfg.classes,
+                    data=sub_data,
+                    fraction=sub_fraction,
+                )
+            )
+        if len(datasets_list) > 1:
+            return YOLOConcatDataset(datasets_list)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py:306]: ⚠️ **HIGH**: `YOLOConcatDataset` does not expose a `labels` attribute, but the standard detection trainer unconditionally accesses `self.train_loader.dataset.labels` when `plots=True` (the default) in `plot_training_labels()`. This makes normal multi-dataset training fail during setup with `AttributeError`; it also breaks auto-batch and nonzero `cls_pw` paths that access the same attribute. Expose a combined labels view (and any other dataset metadata required by these trainer paths) on `YOLOConcatDataset`.

File: ultralytics/data/utils.py [hunk 2]
```diff
@@ -483,6 +483,53 @@ def check_det_dataset(dataset: str, autodownload: bool = True, split: str = "")
...
+
+        sub_dicts = []
+        for ds in data["datasets"]:
+            if isinstance(ds, dict):
+                ds_yaml = ds.get("data") or ds.get("path")
+                if not ds_yaml:
+                    raise ValueError(
+                        f"Each dataset entry dictionary in '{dataset}' must contain a 'data' or 'path' key. Got: {ds}"
+                    )
+                sub = check_det_dataset(str(ds_yaml), autodownload, split)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/utils.py:501]: 💡 **MEDIUM**: Sub-dataset YAML references are resolved relative to the process working directory (or repository search paths), not relative to the parent multi-dataset YAML (`file`). Consequently, a portable config such as `configs/combined.yaml` containing `datasets: [dataset_a.yaml]` fails when launched outside `configs/`. Resolve local relative entries against `file.parent` before recursively calling `check_det_dataset` (in both the dictionary and string entry branches).

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The semantic remapping flow and cache-preserving label handling are well covered by focused tests. Two integration issues remain: concatenated datasets do not meet the trainer's expected dataset interface, and relative sub-dataset YAML paths are not anchored to the combined configuration file.

💬 Posted 2 inline comments

**Comment by MohanadMahran** [comment]
I have read the CLA Document and I sign the CLA.

**Comment by glenn-jocher** [comment]
Thank you for the thoughtful contribution. I am closing this version because it is not minimal or production-ready under the repository Core Principles: it adds 932 lines and a new registry, concat-dataset abstraction, schema, documentation surface, and 466-line test suite while deleting one line. The standard trainer still expects dataset attributes such as `labels`, so normal training fails, and relative child YAML paths are resolved from the process directory rather than the parent config. Fixing those findings would expand the new architecture further rather than simplify ownership. Please start with a design issue that settles the multi-dataset contract, then send smaller staged PRs with a real end-to-end training path and a literal `Deleted:` line. Thanks again for exploring this.
